### PR TITLE
Refactor http source functionality for supporting a new OpenSearch API source in DataPrepper

### LINKED ISSUE #248: OpenSearch Bulk API Source
## Summary

This creates a new Data Prepper source which accepts data in the form of the [OpenSearch Bulk API](https://opensearch.org/docs/latest/api-reference/document-apis/bulk/).

## Configuration

```
source:
  opensearch_api:
    port: 9200
    path_prefix: opensearch/
```

## Operations

The `_bulk` API supports:

* `index`
* `create`
* `update`
* `delete`

This source can do something similar to what the `dynamodb` source does. Specifically it should include the `opensearch_action` metadata.

## Sample

```
POST opensearch/_bulk
{ "index": { "_index": "movies", "_id": "tt1979320" } }
{ "title": "Rush", "year": 2013 }
```

The above request is the simplest case since it is an `index` request.

It creates an Event with data such as:

```
{ "_id": "tt1979320" "title": "Rush", "year": 2013 }
```

Additionally, the event will need metadata that we can use in the `opensearch` sink.

```
opensearch_action: "index"
opensearch_index: "movies"
opensearch_id: "tt1979320"
```

## Query parameters

The `_bulk` API supports a few query parameters. The source should also support most of these and provide some of them as metadata.

* `pipeline` -> Sets metadata: `opensearch_pipeline`
* `routing` -> Sets metadata: `opensearch_routing`
* `timeout` -> Configures an alternate timeout for the request in the source. This probably doesn't need to be provided downstream.

Some other parameters that we may wish to support:

* `refresh`
* `require_alias`
* `wait_for_active_shards`

Finally, we should not support these parameters as they are being deprecated.

* `type`

## Response

Being able to provide the `_bulk` API response may be more challenging. There are a few reasons:

1. Unless end-to-end acknowledgments are enabled, we won't have any knowledge of the writes.
2. Even when acknowledgments are enabled all the metadata needed in a typical response is still not available.

An initial version could provide responses that either have empty values (where appropriate) or use synthetic values.

### PR DESCRIPTION
**Pull request by sb2k16**
### Description

In order for DataPrepper to support all OpenSearch Document API(s), we need to build a new source similar to the existing `http` source. As a first step, this pull request is intended to refactor some of the existing functionalities in `http` source including ssl support, certificate handling, throttling functionality and request rejection handling to a new `http-source-common` package under `data-prepper-plugins`. 
 
### Issues Resolved
Contributes to #248
 
### Check List
- [X] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
- [X] New functionality has javadoc added
- [X] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/http-source/src/test/java/org/opensearch/dataprepper/plugins/source/loghttp/HTTPSourceConfigTest.java [hunk 1]
```diff
@@ -278,17 +75,13 @@ void isPathValid_should_return_false_for_invalid_path() throws NoSuchFieldExcept
...
-                field.setAccessible(true);
-                field.set(httpSourceConfig, value);
-            } finally {
-                field.setAccessible(false);
-            }
-        }
-
-        private String getS3FilePath() {
-            return S3_PREFIX.concat(UUID.randomUUID().toString());
+    private void reflectivelySetField(final HTTPSourceConfig httpSourceConfig, final String fieldName, final Object value) throws NoSuchFieldException, IllegalAccessException {
```
Comment (dlvenable) [inline data-prepper-plugins/http-source/src/test/java/org/opensearch/dataprepper/plugins/source/loghttp/HTTPSourceConfigTest.java:78]: This was added as a test helper. Let's use that:

https://github.com/opensearch-project/data-prepper/blob/6648d9fea792cfca889c69f62ce50af6ee12392a/data-prepper-test-common/src/main/java/org/opensearch/dataprepper/test/helper/ReflectivelySetField.java#L29-L38

File: data-prepper-plugins/http-source-common/src/main/java/org/opensearch/dataprepper/http/HttpServerConfig.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+import org.opensearch.dataprepper.model.configuration.PluginModel;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.micrometer.core.instrument.util.StringUtils;
+import jakarta.validation.constraints.AssertTrue;
+import jakarta.validation.constraints.Min;
+
+/**
+ * HttpServerConfig class holds the common http configurations defined in http source.
+*/
+public class HttpServerConfig {
```
Comment (dlvenable) [inline data-prepper-plugins/http-source-common/src/main/java/org/opensearch/dataprepper/http/HttpServerConfig.java:18]: Let's make this an interface. Then you can create a `BaseHttpServerConfig` that implements the interface and has the same implementation that you have here.

We may find that we want a different set of default values or even different configurations between sources. So decoupling the usage of this interface from the concrete configuration will allow these to vary as needed.

**Review by dlvenable — CHANGES_REQUESTED** [review]
Thank you @sb2k16 for this great contribution! I have a few comments, but otherwise it looks good.

File: data-prepper-plugins/http-source/src/main/java/org/opensearch/dataprepper/plugins/source/loghttp/HTTPSourceConfig.java [hunk 3]
```diff
@@ -154,75 +43,7 @@ public String getPath() {
...
-
-    public boolean hasHealthCheckService() {
-        return healthCheckService;
-    }
-
-    public boolean isUnauthenticatedHealthCheck() {
-        return unauthenticatedHealthCheck;
-    }
-
     public CompressionOption getCompression() {
```
Comment (dlvenable) [inline data-prepper-plugins/http-source/src/main/java/org/opensearch/dataprepper/plugins/source/loghttp/HTTPSourceConfig.java:46]: Do we need to add this to `HttpServerConfig`? How will it be used if it is only here?

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/http-source-common/build.gradle [hunk 4]
```diff
@@ -9,6 +9,7 @@ plugins {
 
 dependencies {
     implementation project(':data-prepper-plugins:common')
+    implementation project(':data-prepper-test-common')
```
Comment (dlvenable) [inline data-prepper-plugins/http-source-common/build.gradle:12]: Please make this `testImplementation` and move toward the bottom to sit with the other `testImplementation` dependencies.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/http-source/build.gradle [hunk 5]
```diff
@@ -10,8 +10,10 @@ plugins {
 dependencies {
     implementation project(':data-prepper-api')
     implementation project(':data-prepper-plugins:blocking-buffer')
+    implementation project(':data-prepper-plugins:http-source-common')
     implementation project(':data-prepper-plugins:common')
     implementation project(':data-prepper-plugins:armeria-common')
+    implementation project(':data-prepper-test-common')
```
Comment (dlvenable) [inline data-prepper-plugins/http-source/build.gradle:16]: Please make this `testImplementation` and move to set with the other `testImplementation` dependencies a few lines below.

**Review by dlvenable — CHANGES_REQUESTED** [review]

**Review by kkondaka — DISMISSED** [review]

**Review by dlvenable — APPROVED** [review]

**Review by kkondaka — APPROVED** [review]

**Pull request merged**